Concurrent downloads & installation

djc opened the pull request:
This is what I had in mind. It seems to compile but I've done no testing or benchmarking.

@FranciscoTGouveia maybe that's something you can have a look at?

Alternative to

- #4471 

Obsoletes

- #4567

djc committed:
dist: move URL alteration logic into DownloadCfg method

djc committed:
dist: install while downloading

FranciscoTGouveia commented:
It compiles, but I believe there is an infinite loop; at least, I could not get any toolchain installation to finish as it seemed to hang on the download of the last component.

Additionally, I would like to note that when a component is being installed, although ongoing downloads continue normally, new component downloads cannot start.
As a result, during an installation (assuming `RUSTUP_CONCURRENT_DOWNLOADS=2`, which is the default), no more than one download is active at a time.

This isn’t an issue when `RUSTUP_CONCURRENT_DOWNLOADS` is configured to maximize concurrency, but otherwise, it can have a noticeable impact on performance (I wasn’t able to benchmark this due to the infinite loop).

Finally, I would like to point out that running both downloads and installations on the same thread prevents the progress bars for other downloads from updating (see [here](https://github.com/rust-lang/rustup/pull/4567#issuecomment-3448708164)).

FranciscoTGouveia commented on src/dist/manifestation.rs:
```diff
@@ -255,15 +214,40 @@ impl Manifestation {
+        let mut component_iter = components.iter();
+        let mut cleanup_downloads = vec![];
+        loop {
+            while downloads.len() < concurrent_downloads {
```
This loop will run _ad aeternum_.

I believe the issue is that after the penultimate component finishes, `downloads.len()` remains 1.
Since the default value for `concurrent_downloads` is 2, the condition is never satisfied, and the loop never exits.

Even if `RUSTUP_CONCURRENT_DOWNLOADS` is overridden to `1`, the issue persists; the last component will complete its download and installation, but the loop will still not exit.

djc replied:
Thanks for investigating, will fix this.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -242,15 +214,40 @@ impl Manifestation {
+        let mut downloads = FuturesUnordered::new();
+        let mut component_iter = components.iter();
+        let mut cleanup_downloads = vec![];
+        loop {
```
@djc Thanks for the testbed!

After careful comparisons, I think this is equivalent to [`b5e9e4d` (#4471)](https://github.com/rust-lang/rustup/pull/4471/commits/b5e9e4d7559a53351c6ad7d4bba5caddf204f1fb) once we squash [`5800cba` (#4471)](https://github.com/rust-lang/rustup/pull/4471/commits/5800cba76419fcdee5100086b0d3a005b99ec410) back onto it to eliminate the use of channels, modulo some edge cases such as empty todo list and download failures.

The core of this approach includes directly issuing a blocking call to what we have pulled from a `FuturesUnordered`. When that blocking call is running, there will be no progress updates and no new installations being initiated. So this is the same as what we've had after the very first step of #4471.

djc replied:
Thanks for looking into. Do you think it's useful to merge (something like) this and proceed from there?

> When that blocking call is running, there will be no progress updates and no new installations being initiated.

If this is the main issue that's left, it feels like there might be easier (and better) ways to solve this.

rami3l replied:
@djc I think that is THE remaining issue, yes.

However I still have concerns regarding the aforementioned edge cases that I and @FranciscoTGouveia have found out when working on #4471 so more careful checks have to be carried out.

I think it'd be nice if both of us can have another look at this PR to ensure the semantics are right, it might take some time though while I try to coordinate 🙇

djc replied:
> However I still have concerns regarding the aforementioned edge cases that I and @FranciscoTGouveia have found out when working on #4471 so more careful checks have to be carried out.
> 
> I think it'd be nice if both of us can have another look at this PR to ensure the semantics are right, it might take some time though while I try to coordinate 🙇

Sounds great!

djc committed:
dist: yield self when download is complete

djc committed:
dist: store more context in ComponentBinary

djc commented:
I've fixed this up so that it can at least pass basic manual tests. CI is failing because I changed the log output -- happy to fix this up if we decide that we want to move forward with this.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -242,15 +216,41 @@ impl Manifestation {
+                Some(Ok((bin, downloaded))) => (bin, downloaded),
+                Some(Err(e)) => return Err(e),
+                None => {
+                    if let Some(bin) = component_iter.next() {
```
This seems inconsistent with the `while` section below in the number of tasks being added to the todo list.

djc replied:
Fair point. I've tweaked the code to improve on this.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -240,15 +214,38 @@ impl Manifestation {
+                break;
+            }
+
+            let installable = match downloads.next().await {
```
Remark: I think this is the same as `let installable = downloads.next().await.transpose()?;`.

djc replied:
Nice, fixed!

rami3l reviewed (approved):
I think this works out pretty well during both code inspection and manual tests, so I consider this to be a somewhat faithful reproduction of the first step of #4471. Given that, I agree with merging this despite the fact that the project will suffer from temporary UX regressions.

I'd like to invite @FranciscoTGouveia do another round of review from his side to make sure everything is retained from the original solution.

djc commented:
Here's what I think we could do next to solve the remaining UX problem and even further improve the UX:

- Update the progress indicator during unpacking, using a [`ProgressBarIter`](https://docs.rs/indicatif/latest/indicatif/struct.ProgressBarIter.html) wrapper
- Update the progress indicator during installing, by making installation "async", maybe by calling [`yield_now()`](https://docs.rs/tokio/latest/tokio/task/fn.yield_now.html) every 250ms or so from `DirectoryPackage::install()`

(A more principled approach might be to use `tokio::fs` but because there's a lot of layers in our install process that's probably a bunch of work, if not exactly hard.)
